Implement the `Mapping` interface for `GroupingLatestValueCache`

shsms opened the pull request.

shsms committed:
Implement `Mapping` for `GroupingLatestValueCache`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update release notes

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on src/frequenz/channels/experimental/_grouping_latest_value_cache.py:
```diff
@@ -39,20 +39,21 @@
 
 T_co = typing.TypeVar("T_co", covariant=True)
+T = typing.TypeVar("T")
 HashableT = typing.TypeVar("HashableT", bound=typing.Hashable)
```
Optional: Document these (adding a docstring below them), so they are shown in the docs index.

shsms replied:
They are internal, not exposed from the `_grouping_latest_value_cache` module.  Not sure if they should be though.

llucax replied:
I guess both are fine. If they are not exported, the docs will show it as a "string" without any links, and it will not be present in the ToC. If the types are self explanatory we can keep it as it is. If they are not documented/exported, users will not know about the possible `bound=` arguments or *constraints* (or if it is co/contravariant, but for those we usually use the suffixes), so if it is important to the user to know any of that info, that could be a good reason to export and document.

shsms replied:
I've added a commit to do this.  I'm somewhat doubtful about exposing them from the `experimental` package directly, but that seems to be the best option available.

llucax commented on src/frequenz/channels/experimental/_grouping_latest_value_cache.py:
```diff
@@ -39,20 +39,21 @@
-    as long as there has been one value received.
-    """
+class GroupingLatestValueCache(
+    typing.Generic[T_co, HashableT], Mapping[HashableT, T_co]
```
It shouldn't be necessary to keep the `typing.Generic[T_co, HashableT]` AFAIK, you have them indirectly via `Mapping[HashableT, T_co]`.

shsms replied:
done

llucax reviewed (commented):
Not sure if it is worth keeping `c.has_key(k)` instead of requiring the use of `k in c`.

Also not sure if it might be worth implementing also the methods automatically mixed in by `Mapping` and its subclasses: `__contains__`, `items`, `values`, `__eq_`_, and `__ne__`, otherwise they are probably calculated via the iterator or something like that, which could be pretty inefficient.

![image](https://github.com/user-attachments/assets/888583f2-1dd5-496e-8028-ac4210617fca)
![image](https://github.com/user-attachments/assets/e385ff32-e6f9-4eb9-9e80-8a87a45e48be)

shsms commented:
> Not sure if it is worth keeping `c.has_key(k)` instead of requiring the use of `k in c`.
> 
> Also not sure if it might be worth implementing also the methods automatically mixed in by `Mapping` and its subclasses: `__contains__`, `items`, `values`, `__eq_`_, and `__ne__`, otherwise they are probably calculated via the iterator or something like that, which could be pretty inefficient.

All done.

shsms committed:
Expose `TypeVar`s used to `GroupingLatestValueCache`

Making the part of the public API makes it easy for users to look them
up and understand how to use them.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>
